diff --git a/R/parse_data.R b/R/parse_data.R
index 612200e..a3c7232 100644
--- a/R/parse_data.R
+++ b/R/parse_data.R
@@ -137,5 +137,5 @@ clean_parse_data <- function() {
 
 # Needed to work around https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=16756
 get_tokens <- function(srcref) {
-  getParseData(srcref) %||% get_parse_data(attr(getSrcref(srcref), "srcfile"))
+  get_parse_data(attr(getSrcref(srcref), "srcfile"))
 }
diff --git a/R/utils.R b/R/utils.R
index ebb151b..3cfa2cd 100644
--- a/R/utils.R
+++ b/R/utils.R
@@ -92,7 +92,12 @@ split_on_line_directives <- function(lines) {
   directive_lines <- which(!is.na(matches$line_number))
   file_starts <- directive_lines + 1
   file_ends <- c(directive_lines[-1] - 1, length(lines))
-  res <- mapply(function(start, end) lines[start:end], file_starts, file_ends)
+  res <- mapply(
+    function(start, end) lines[start:end],
+    file_starts,
+    file_ends,
+    SIMPLIFY = FALSE
+  )
   names(res) <- na.omit(matches$filename)
   res
 }
